src/osd/scrubber/scrub_machine.h

```diff
@@ -776,16 +839,75 @@ struct ReplicaActive : sc::state<
       const ReplicaReserveReq&,
       bool async_request);
 
+  /**
+   * the queued reservation request was granted by the async reserver.
+   * Notify the Primary.
+   * Returns 'false' if the reservation is not the last one to be received
+   * by this replica.
+   */
+  bool granted_by_reserver(const AsyncScrubResData& resevation);
+
   /// handle a 'release' from a primary
   void on_release(const ReplicaRelease& ev);
 
-  /// cancel the reserver request.
-  /// The 'failure' re 'log_failure' is logged if we are not reserved to
-  /// begin with.
+  /**
+   * cancel the reserver request.
+   * The 'failure' re 'log_failure' is logged if we are not reserved to
+   * begin with.
+   */
   void clear_reservation_by_remote_primary(bool log_failure);
 
+  /**
+   * cancel_ungranted_request - cancel the reservation request if it was not
+   * already granted by the async reserver.
+   */
+  void cancel_ungranted_request();
```

Comment:
cancel_ungranted_request and clear_reservation_by_remote_primary are very nearly the same thing.  I don't really understand why they are distinct.

Reply:
cancel_ungranted_request() cancels just requests that were not granted, allowing
me to call it unconditionally at ReplicaWaitingReservation dtor.